GenericJSON/Merging.swift

```diff
@@ -36,6 +37,21 @@ extension JSON {
             merged[key] = lhs[key]?.merging(with: rhs[key]!)
         }
 
-        return JSON.object(merged)
+        self = .object(merged)
+    }
+
+    /// Return a new JSON value by merging two other ones
+    ///
+    /// This is a non-mutating version of `merge(with:)`.
+    public func merging(with new: JSON) -> JSON {
+        return with(self) {
+            $0.merge(with: new)
+        }
     }
 }
+
+private func with<T>(_ item: T, update: (inout T) throws -> Void) rethrows -> T {
```

Comment:
```suggestion
private func copying<T>(_ item: T, update: (inout T) throws -> Void) rethrows -> T {
```
I really like this snippet, however i think it should not be called with(self), as it might suggests an operation applied on the current instance. Since we are creating a copy of the item, i'm proposing a different name that explicits that.
Sorry for being pedantic :)
Alternatively:
```suggestion
private func with<T>(copyOf item: T, update: (inout T) throws -> Void) rethrows -> T {
```